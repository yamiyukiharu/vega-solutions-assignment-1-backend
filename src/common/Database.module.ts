import { Module } from "@nestjs/common";
import { TypeOrmModule } from "@nestjs/typeorm";
import { AppConfigModule } from "./AppConfig.module";
import { ConfigService } from "@nestjs/config";
import { Transaction } from "src/transaction/models/Transaction.model";

@Module({
  imports: [
    TypeOrmModule.forRootAsync({
      imports: [AppConfigModule],
      useFactory: (configService: ConfigService) => ({
        type: 'mongodb',
        url: configService.get('DB_URL'),
        entities: [Transaction],
        synchronize: true,
        useNewUrlParser: true,
        logging: true,
      }),
      inject: [ConfigService],
    }),
  ],
  exports: [TypeOrmModule]
})
export class DatabaseModule { }