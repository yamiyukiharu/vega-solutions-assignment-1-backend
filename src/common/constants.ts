export const REPORTS_QUEUE = 'transactions-report';

export const RECORD_QUEUE = 'transactions-record';