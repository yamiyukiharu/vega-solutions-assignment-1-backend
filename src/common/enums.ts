export enum Protocol {
  UNISWAPV3 = 'uniswapv3',
}

export enum Pool {
  ETH_USDC = 'usdc_eth',
}

export enum ReportStatus {
  PENDING = 'pending',
  IN_PROGRESS = 'in_progress',
  COMPLETED = 'completed',
  FAILED = 'failed',
}