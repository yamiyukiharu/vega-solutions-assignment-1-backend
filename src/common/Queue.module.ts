
import { BullModule } from '@nestjs/bull';
import { Module } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { AppConfigModule } from './AppConfig.module';

@Module({
  imports: [
    BullModule.registerQueueAsync({
      imports: [AppConfigModule],
      name: 'transactions-report',
      useFactory: async (configService: ConfigService) => ({
        redis: {
          host: configService.get('QUEUE_HOST'),
          port: configService.get('QUEUE_PORT'),
        },
      }),
      inject: [ConfigService],
    }),
  ],
  exports: [BullModule],
})
export class QueueModule { }
