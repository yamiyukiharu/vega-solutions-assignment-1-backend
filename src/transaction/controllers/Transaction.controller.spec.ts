import * as request from 'supertest';
import { Test, TestingModule } from '@nestjs/testing';
import { TransactionController } from './Transaction.controller';
import { TransactionService } from '../services/Transaction.service';
import {
  GenerateReportRequest,
} from '../dtos/transaction.dto';
import {
  INestApplication,
} from '@nestjs/common';
import { Pool, Protocol, ReportStatus } from 'src/common/enums';

describe('TransactionController', () => {
  let app: INestApplication;
  let controller: TransactionController;
  let serviceMock = {
    getTransactionList: jest.fn(),
    getTransactionCount: jest.fn(),
    getTransactionByHash: jest.fn(),
    triggerReportGeneration: jest.fn(),
    getReportStatus: jest.fn(),
    getReport: jest.fn(),
  };

  beforeEach(async () => {
    // Create a new testing module with the TransactionController and a mocked TransactionService
    const module: TestingModule = await Test.createTestingModule({
      controllers: [TransactionController],
      providers: [
        {
          provide: TransactionService,
          useValue: serviceMock,
        },
      ],
    }).compile();

    // Get the TransactionController instance and the mocked TransactionService instance
    controller = module.get<TransactionController>(TransactionController);

    app = module.createNestApplication();
    await app.init();
  });

  describe('getTransactions', () => {
    it('should return a list of transactions', async () => {
      // Arrange
      const transactions = [{ id: 1 }, { id: 2 }];
      serviceMock.getTransactionList.mockResolvedValue(transactions);
      serviceMock.getTransactionCount.mockResolvedValue(transactions.length);

      const protocol = Protocol.UNISWAPV3;
      const pool = Pool.ETH_USDC;
      const page = 1;
      const limit = 10;

      // Act
      const result = await request(app.getHttpServer())
        .get(
          `/v1/transactions?protocol=${Protocol.UNISWAPV3}&pool=${Pool.ETH_USDC}&page=${page}&limit=${limit}`,
        )
        .expect(200);

      // Assert
      expect(serviceMock.getTransactionList).toHaveBeenCalled();

      expect(result.body).toMatchSnapshot();
    });

    it('should return a single transaction when a hash is provided', async () => {
      // Arrange
      const transaction = { id: '0x1234' };
      serviceMock.getTransactionByHash.mockResolvedValue(transaction);
      serviceMock.getTransactionCount.mockResolvedValue(1);

      const protocol = Protocol.UNISWAPV3;
      const pool = Pool.ETH_USDC;

      // Act
      const result = await request(app.getHttpServer())
        .get(
          `/v1/transactions?protocol=${Protocol.UNISWAPV3}&pool=${Pool.ETH_USDC}&hash=${transaction.id}`,
        )
        .expect(200);

      // Assert
      expect(serviceMock.getTransactionByHash).toBeCalled();

      expect(result.body).toMatchSnapshot();
    });

    it('should throw a NotFoundException when a hash is provided but no transaction is found', async () => {
      // Arrange
      const transaction = { id: '0x1234' };
      serviceMock.getTransactionByHash.mockResolvedValue(null);
      serviceMock.getTransactionCount.mockResolvedValue(1);

      const protocol = Protocol.UNISWAPV3;
      const pool = Pool.ETH_USDC;

      // Act
      const result = await request(app.getHttpServer())
        .get(
          `/v1/transactions?protocol=${Protocol.UNISWAPV3}&pool=${Pool.ETH_USDC}&hash=${transaction.id}`,
        )
        .expect(404);

      // Assert
      expect(serviceMock.getTransactionByHash).toBeCalled();

      expect(result.body).toMatchSnapshot();
    });
  });

  describe('generateReport', () => {
    it('should trigger report generation and return a Location header', async () => {
      // Arrange
      const reportId = 'report-id';
      serviceMock.triggerReportGeneration.mockResolvedValue(reportId);

      const body: GenerateReportRequest = {
        protocol: Protocol.UNISWAPV3,
        pool: Pool.ETH_USDC,
        startTime: '2022-01-01T00:00:00.000Z',
        endTime: '2022-01-02T00:00:00.000Z',
      };

      // Act
      const result = await request(app.getHttpServer())
        .post(`/v1/transactions/reports`)
        .send(body)
        .expect(202);

      // Assert
      expect(serviceMock.triggerReportGeneration).toHaveBeenCalled();
      expect(result.header.location).toMatchSnapshot();
      expect(result.body).toMatchSnapshot();
    });

    it('should throw an HttpException when report generation fails', async () => {
      // Arrange
      serviceMock.triggerReportGeneration.mockResolvedValue(null);

      const body: GenerateReportRequest = {
        protocol: Protocol.UNISWAPV3,
        pool: Pool.ETH_USDC,
        startTime: '2022-01-01T00:00:00.000Z',
        endTime: '2022-01-02T00:00:00.000Z',
      };

      // Act
      const result = await request(app.getHttpServer())
        .post(`/v1/transactions/reports`)
        .send(body)
        .expect(500);

      // Assert
      expect(serviceMock.triggerReportGeneration).toHaveBeenCalled();
      expect(result.body).toMatchSnapshot();
    });
  });

  describe('getReportStatus', () => {
    it('should return the status of a report', async () => {
      // Mock the TransactionService's getReportStatus method to return a report status
      const reportStatus = 'completed';
      const reportId = 'report-id';
      serviceMock.getReportStatus.mockResolvedValue(reportStatus);

      // Act
      const result = await request(app.getHttpServer())
        .get(`/v1/transactions/reports/status/${reportId}`)
        .expect(200);

      // Assert
      expect(serviceMock.getReportStatus).toHaveBeenCalled();
      expect(result.body).toEqual({
        status: reportStatus,
      });
    });

    it('should throw a NotFoundException when no report is found', async () => {
      // Mock the TransactionService's getReportStatus method to return a report status
      const reportId = 'report-id';
      serviceMock.getReportStatus.mockResolvedValue(null);

      // Act
      const result = await request(app.getHttpServer())
        .get(`/v1/transactions/reports/status/${reportId}`)
        .expect(404);

      // Assert
      expect(serviceMock.getReportStatus).toHaveBeenCalled();
      expect(result.body).toMatchSnapshot();
    });
  });

  describe('getReport', () => {
    it('should return a report', async () => {
      // Arrange
      const reportId = 'report-id';
      const report = {
        id: 'report-id',
        protocol: Protocol.UNISWAPV3,
        pool: Pool.ETH_USDC,
        startTimestamp: 16023123123,
        endTimestamp: 16023193123,
        status: ReportStatus.COMPLETED,
        transactions: [
          {
            hash: '0x1234',
            timestamp: 16023123123,
            fee: '1234',
          },
          {
            hash: '0x1235',
            timestamp: 16023123129,
            fee: '1234',
          },
        ],
        total: 2,
        totalFee: {
          eth: '10000000',
          usdt: '12.02',
        },
      };
      serviceMock.getReport.mockResolvedValue(report);

      // Act
      const result = await request(app.getHttpServer())
        .get(`/v1/transactions/reports/${reportId}`)
        .expect(200);

      // Assert
      expect(serviceMock.getReportStatus).toHaveBeenCalled();
      expect(result.body).toEqual(report);
    });

    it('should throw a NotFoundException when no report is found', async () => {
      // Arrange
      serviceMock.getReport.mockResolvedValue(null);

      // Act
      const result = await request(app.getHttpServer())
        .get(`/v1/transactions/reports/report-id`)
        .expect(404);

      // Assert
      expect(serviceMock.getReportStatus).toHaveBeenCalled();
      expect(result.body).toMatchSnapshot();
    });
  });
});
