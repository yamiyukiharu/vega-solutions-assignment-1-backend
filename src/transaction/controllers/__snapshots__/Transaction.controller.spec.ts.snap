// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionController generateReport should throw an HttpException when report generation fails 1`] = `
{
  "message": "Failed to trigger report generation",
  "statusCode": 500,
}
`;

exports[`TransactionController generateReport should trigger report generation and return a Location header 1`] = `"v1/transactions/reports/status/report-id"`;

exports[`TransactionController generateReport should trigger report generation and return a Location header 2`] = `
{
  "location": "v1/transactions/reports/status/report-id",
}
`;

exports[`TransactionController getReport should throw a NotFoundException when no report is found 1`] = `
{
  "error": "Not Found",
  "message": "Report with id report-id not found",
  "statusCode": 404,
}
`;

exports[`TransactionController getReportStatus should throw a NotFoundException when no report is found 1`] = `
{
  "error": "Not Found",
  "message": "Report with id report-id not found",
  "statusCode": 404,
}
`;

exports[`TransactionController getTransactions should return a list of transactions 1`] = `
{
  "data": [
    {
      "id": 1,
    },
    {
      "id": 2,
    },
  ],
  "limit": "10",
  "page": "1",
  "total": 2,
}
`;

exports[`TransactionController getTransactions should return a single transaction when a hash is provided 1`] = `
{
  "data": [
    {
      "id": "0x1234",
    },
  ],
}
`;

exports[`TransactionController getTransactions should throw a NotFoundException when a hash is provided but no transaction is found 1`] = `
{
  "error": "Not Found",
  "message": "Transaction with hash 0x1234 not found",
  "statusCode": 404,
}
`;
